build/utils/npm.go

```diff
@@ -9,189 +9,284 @@ import (
+	return dependenciesMap, jsonparser.ObjectEach(data, func(key []byte, value []byte, dataType jsonparser.ValueType, offset int) (err error) {
 		if string(key) == "dependencies" {
-			err = parseDependencies(value, typeRestriction, []string{moduleId}, results, log)
+			err = parseDependencies(value, []string{moduleId}, dependenciesMap, cacache, parseFunc, log)
 		}
 		return err
 	})
 }
 
+func GetNpmVersion(executablePath string, log utils.Log) (*version.Version, error) {
+	versionData, _, err := RunNpmCmd(executablePath, "", Version, nil, log)
+	if err != nil {
+		return nil, err
+	}
+	return version.NewVersion(string(versionData)), nil
+}
+
+// A partial struct of the dependency json object coming from npm ls output.
+type npmLsDependency struct {
+	Name      string
+	Version   string
+	Integrity string
+	InBundle  bool
+	Dev       bool
+	// Missing peer dependency in npm version 7/8
+	Missing bool
+	// Problems with missing peer dependency in npm version 7/8
+	Problems []string
+	// Missing  peer dependency in npm version 6
+	// Bound to 'legacyNpmLsDependency' struct
+	PeerMissing []*peerMissing
+}
+
+type legacyNpmLsDependency struct {
+	Name        string
+	Version     string
+	Integrity   string `json:"_integrity,omitempty"`
+	InBundle    bool   `json:"_inBundle,omitempty"`
+	Dev         bool   `json:"_development,omitempty"`
+	PeerMissing []*peerMissing
+}
+
+type peerMissing struct {
+	RequiredBy string
+	Requires   string
+}
+
+func (lnld *legacyNpmLsDependency) toNpmLsDependency() *npmLsDependency {
+	return &npmLsDependency{
+		Name:        lnld.Name,
+		Version:     lnld.Version,
+		Integrity:   lnld.Integrity,
+		InBundle:    lnld.InBundle,
+		Dev:         lnld.Dev,
+		PeerMissing: lnld.PeerMissing,
+	}
+}
+
+// Return name:version of a dependency
+func (nld *npmLsDependency) id() string {
+	return nld.Name + ":" + nld.Version
+}
+
+func (nld *npmLsDependency) getScopes() (scopes []string) {
+	if nld.Dev {
+		scopes = append(scopes, "dev")
+	} else {
+		scopes = append(scopes, "prod")
+	}
+	if strings.HasPrefix(nld.Name, "@") {
+		splitValues := strings.Split(nld.Name, "/")
+		if len(splitValues) > 2 {
+			scopes = append(scopes, splitValues[0])
+		}
+	}
+	return
+}
+
 // Parses npm dependencies recursively and adds the collected dependencies to the given dependencies map.
-func parseDependencies(data []byte, scope string, pathToRoot []string, dependencies *map[string]*entities.Dependency, log utils.Log) error {
+func parseDependencies(data []byte, pathToRoot []string, dependencies map[string]*prototypeNode, cacache *cacache, parseFunc func(data []byte) (*npmLsDependency, error), log utils.Log) error {
 	return jsonparser.ObjectEach(data, func(key []byte, value []byte, dataType jsonparser.ValueType, offset int) error {
-		depName := string(key)
-		ver, _, _, err := jsonparser.Get(data, depName, "version")
-		if err != nil && err != jsonparser.KeyPathNotFoundError {
+		if string(value) == "{}" {
```

Comment:
I'm not sure about this one:
```suggestion
		if dataType == jsonparser.NotExist {
```